### DISCUSSION TITLE
"explain", a theme showing the meaning of each section

### DISCUSSION THREAD

**Discussion by nojhan** [liquidprompt/liquidprompt#854] [Show and tell]
*[lp-explain](https://github.com/nojhan/lp-explain)* is a theme for Liquid Prompt.

It aims at showing the meaning of each section of the default prompt theme.
To do so it displays:

- a box around each section,
- with the name of the corresponding variable in the default theme's template,
- and it adds a hyperlink to it, so that the user can click and rapidly jump to the documentation.

It is a way to answer a feature request I got on the [lp-jolly theme](https://github.com/nojhan/lp-jolly/issues/6).

Ultimately, maybe it would be interesting to ship it as an included theme, and/or a `lp_explain` command to display it? I am not sure.

**Comment by Rycieos** [comment]
First, very cool. Visually clear, links to the docs are fantastic UX, and by implementing it as a theme the user's real data is displayed.

Second, I have a couple of suggestions:
1. Mention in the docs that [`LP_ENABLE_HYPERLINKS`](https://liquidprompt.readthedocs.io/en/stable/config.html#LP_ENABLE_HYPERLINKS) must be enabled for the docs links to work, which is it not by default.
2. My font does not support `🮠`, `🮥`, or `🮠` (Box Drawings Light Diagonal Upper Centre to Middle Left / to Middle Right to Lower Centre / to Middle Left), meaning there will probably be others that are missing those as well. No need to remove it, but having a mention in the docs of some fonts that do support it would be nice. Better yet would be updating Liquid Prompt's suggested fonts as well.

Third, seeing this makes me want to support this in the engine. I am imagining a way for themes to create sections and register them with IDs, and when the engine goes in to "explain" mode, it does this blocking for the theme automatically. However, Liquid Prompt as it currently exists just can't do this automatically. Themes currently have way too much control (which can be a good thing), but means there is no way to support this automatically for all themes. That said, in a next gen prompt engine (like MOST if it ever gets going, see end of #786), this should be a core feature.

So if we can't ship it as a core feature that works for all themes, what can we do? I think your suggestions are good. A command to invoke it would likely be confusing, as it wouldn't work with any other themes.

So here is what I would suggest: as it currently exists, as a theme, it is good enough to be merged into the `themes/` directory. Feel free to make a PR doing that (with docs of course). But, since you are our most prolific theme writer, I can see you wanting your themes to get this treatment as well. If you can think of a good way for themes to "register" their sections with the engine to allow for displaying this info (since currently the themes have all the control), then add a way for that to happen into the engine. I would be happy to help resolve issues or even do some design work on the API for that.

**Reply by nojhan** [reply]
Thanks for the feedback, which is very interesting and useful, as usual :-)

I am not sure that I would want my themes shipped with the core distribution, as I am well aware of the maintenance costs it carries, most of the time for either very few benefits (Neon, Solid, Pureship...), or at the cost of a steep learning curve (Jolly, GitCrux...).

However, having something as an explanation tool is a good improvement of the user experience, and I agree with your analysis that having it as a theme is a good compromise in the current state.

I'll do a PR.

As for MOST, I really like the idea (you should consider finding a name with a better discoverability, call it "plasma prompt" and host it in the liquidprompt organization!), but I'm still trying to find time for learning rust :-)

**Comment by nojhan** [comment]
Proposed as PR #856
